From: Sebastien Guiriec <s-guiriec@ti.com>
Date: Sat, 22 Jan 2011 15:31:55 +0000 (+0100)
Subject: ASoC: McPDM: Update the control of the channels for SW reset
X-Git-Url: http://git.omapzoom.org/?p=kernel%2Fomap.git;a=commitdiff_plain;h=85530db290a41ced7bc00ff1b15ec80bdd102fcf;hp=99b2211d319451dd6ea4bb747bf75b034ce4fe9d

ASoC: McPDM: Update the control of the channels for SW reset

McPDM channels management need to be control in a specific way
in order to reset the internal digital filter of the IP for
Uplink and downlink channels.
Perform a Software reset when we are requestig the module.

Change-Id: Ie5d351604d4162eafca0b4ed2ea0d557d6e90998
Signed-off-by: Sebastien Guiriec <s-guiriec@ti.com>
---

Index: android-2.6.35/sound/soc/omap/omap-mcpdm.c
===================================================================
--- android-2.6.35.orig/sound/soc/omap/omap-mcpdm.c	2011-01-25 11:42:12.457811944 +0530
+++ android-2.6.35/sound/soc/omap/omap-mcpdm.c	2011-01-25 11:45:02.137811587 +0530
@@ -211,12 +211,21 @@
 {
 	int ctrl = omap_mcpdm_read(mcpdm, MCPDM_CTRL);
 
-	if (stream)
+	if (stream) {
+		ctrl |= SW_UP_RST;
+		omap_mcpdm_write(mcpdm, MCPDM_CTRL, ctrl);
 		ctrl |= mcpdm->up_channels;
-	else
+		omap_mcpdm_write(mcpdm, MCPDM_CTRL, ctrl);
+		ctrl &= ~SW_UP_RST;
+		omap_mcpdm_write(mcpdm, MCPDM_CTRL, ctrl);
+	} else {
+		ctrl |= SW_DN_RST;
+		omap_mcpdm_write(mcpdm, MCPDM_CTRL, ctrl);
 		ctrl |= mcpdm->dn_channels;
-
-	omap_mcpdm_write(mcpdm, MCPDM_CTRL, ctrl);
+		omap_mcpdm_write(mcpdm, MCPDM_CTRL, ctrl);
+		ctrl &= ~SW_DN_RST;
+		omap_mcpdm_write(mcpdm, MCPDM_CTRL, ctrl);
+	}
 }
 
 /*
@@ -227,12 +236,21 @@
 {
 	int ctrl = omap_mcpdm_read(mcpdm, MCPDM_CTRL);
 
-	if (stream)
+	if (stream) {
+		ctrl |= SW_UP_RST;
+		omap_mcpdm_write(mcpdm, MCPDM_CTRL, ctrl);
 		ctrl &= ~mcpdm->up_channels;
-	else
+		omap_mcpdm_write(mcpdm, MCPDM_CTRL, ctrl);
+		ctrl &= ~SW_UP_RST;
+		omap_mcpdm_write(mcpdm, MCPDM_CTRL, ctrl);
+	} else {
+		ctrl |= SW_DN_RST;
+		omap_mcpdm_write(mcpdm, MCPDM_CTRL, ctrl);
 		ctrl &= ~mcpdm->dn_channels;
-
-	omap_mcpdm_write(mcpdm, MCPDM_CTRL, ctrl);
+		omap_mcpdm_write(mcpdm, MCPDM_CTRL, ctrl);
+		ctrl &= ~SW_DN_RST;
+		omap_mcpdm_write(mcpdm, MCPDM_CTRL, ctrl);
+	}
 }
 
 /*
@@ -404,6 +422,7 @@
 	struct omap_mcpdm_platform_data *pdata;
 	int ret;
 	int ctrl;
+	int attemps = 0;
 
 	pdev = to_platform_device(mcpdm->dev);
 	pdata = pdev->dev.platform_data;
@@ -417,6 +436,18 @@
 	}
 	mcpdm->free = 0;
 
+	/* Perform SW RESET of McPDM IP */
+	ctrl = omap_mcpdm_read(mcpdm, MCPDM_SYSCONFIG);
+	ctrl |= MCPDM_SOFTRESET;
+	omap_mcpdm_write(mcpdm, MCPDM_SYSCONFIG, ctrl);
+	/* Wait completion of SW RESET */
+	while ((omap_mcpdm_read(mcpdm, MCPDM_SYSCONFIG) & MCPDM_SOFTRESET)) {
+		if (attemps++ > 10000) {
+			udelay(10);
+			dev_err(mcpdm->dev, "Could not RESET McPDM\n");
+		}
+	}
+
 	/* Disable lines while request is ongoing */
 	omap_mcpdm_write(mcpdm, MCPDM_CTRL, 0x00);
 
Index: android-2.6.35/sound/soc/omap/omap-mcpdm.h
===================================================================
--- android-2.6.35.orig/sound/soc/omap/omap-mcpdm.h	2011-01-06 13:14:46.000000000 +0530
+++ android-2.6.35/sound/soc/omap/omap-mcpdm.h	2011-01-25 11:45:02.147843781 +0530
@@ -69,6 +69,11 @@
 #define DMA_UP_ENABLE		0x2
 
 /*
+ * MCPDM_SYSCONFIG bit fields
+ */
+#define MCPDM_SOFTRESET		0x1
+
+/*
  * MCPDM_CTRL bit fields
  */
 
